toolbox/swiftsim/generate_splitting_information.py

```diff
@@ -0,0 +1,416 @@
+        Dictionary holding split information (trees, counts, progenitor ID) that
+        has been subdivided into subarrays. Information based on snapshot N.
+
+    Returns
+    -------
+    new_splits : dict
+        Information about which particle ID (key) has split into other particles since the
+        last simulation output (value).
+    '''
+    new_splits = {}
+
+    # Iterate over unique split trees in snapshot N
+    for tree_index, tree_progenitor_ID in enumerate(new_snapshot_data['split_progenitor_ids']):
+
+        # Check whether the current unique tree already existed in snapshot N-1
+        is_new_tree = tree_progenitor_ID not in old_snapshot_data['split_progenitor_ids']
+
+        if is_new_tree:
+
+            # If we have a new tree, all new particle IDs have as their progenitor the
+            # particle ID that originated this unique tree.
+            progenitor_id = tree_progenitor_ID
+
+            new_ids = new_snapshot_data['split_particle_ids'][tree_index]
+            new_ids = new_ids[new_ids != progenitor_id]
+
+            # We could encounter cases where a particle has split and its descendants
+            # have dissapeared from the simulation
+            if len(new_ids) > 0:
+                new_splits[progenitor_id] = new_ids
+
+        else:
+            # Different particles within the tree could have split simultaneously. We need
+            # to be a bit more careful.
+            tree_index_old = np.where(old_snapshot_data['split_progenitor_ids'] == tree_progenitor_ID)[0][0] 
+
+            # Compare the same unique trees between snapshots N and N-1 to see how particles have split
+            new_splits.update(get_splits_of_existing_tree(old_snapshot_data['split_particle_ids'][tree_index_old],
+                                                          old_snapshot_data['split_trees'][tree_index_old],
+                                                          old_snapshot_data['split_counts'][tree_index_old],
+                                                          new_snapshot_data['split_particle_ids'][tree_index],
+                                                          new_snapshot_data['split_trees'][tree_index]))
+
+    return new_splits
+
+def save(split_dictionary, file_path):
+    '''
+    It saves the mapping between split particles in hdf5 files, to
+    be read by HBT+.
+
+    Parameters
+    ----------
+    split_dictionary : dict
+        Information about which particle ID (key) has split into other particles since the
+        last simulation output (value).
+    file_path : str
+        Where to save the HDF5 file containing the map of particle splits.
+    '''
+    # We first need to turn the dictionary into an array used to create a map
+    total_splits = np.array([len(x) for x in split_dictionary.values()]).sum()
+
+    # For completeness purposes, save an empty hdf5 even when we have no splits
+    if(total_splits == 0):
+        with h5py.File(file_path, 'a') as file:
```

Comment:
Is there a reason to open the file in append mode rather than write?

Reply:
Not really. Should I change?

Reply:
Changed